src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java

```diff
@@ -518,29 +528,45 @@ private void detectAndAddExternalContainerInfo(Request request, CommandInfo.Buil
   }
 
   private TaskInfo.Builder getTaskInfoBuilder(Offer offer, Request request, TaskID taskId, CommandInfo.Builder commandBuilder) {
-      return TaskInfo.newBuilder()
+    TaskInfo.Builder builder = TaskInfo.newBuilder()
         .setName("task " + taskId.getValue())
         .setTaskId(taskId)
         .setSlaveId(offer.getSlaveId())
-        .addResources(
+        .setCommand(commandBuilder.build());
+
+    double cpusNeeded = request.request.cpus;
+    double memNeeded = request.request.mem;
+
+    for (Resource r : offer.getResourcesList()) {
+      if (r.getName().equals("cpus") && cpusNeeded > 0) {
+        double cpus = Math.min(r.getScalar().getValue(), cpusNeeded);
+        builder.addResources(
             Resource
                 .newBuilder()
                 .setName("cpus")
                 .setType(Value.Type.SCALAR)
+                .setRole(r.getRole())
                 .setScalar(
                     Value.Scalar.newBuilder()
-                        .setValue(request.request.cpus).build()).build())
-        .addResources(
+                        .setValue(cpus).build()).build());
+        cpusNeeded -= cpus;
+      } else if (r.getName().equals("mem") && memNeeded > 0) {
+        double mem = Math.min(r.getScalar().getValue(), memNeeded);
+        builder.addResources(
             Resource
                 .newBuilder()
                 .setName("mem")
                 .setType(Value.Type.SCALAR)
+                .setRole(r.getRole())
                 .setScalar(
-                    Value.Scalar
-                        .newBuilder()
-                        .setValue((1 + JVM_MEM_OVERHEAD_FACTOR) * request.request.mem)
```

Comment:
Thanks for the hint. I missed this change when rebaseing.
I just pushed a new version including the jvm overhead factor.